pymeasure/adapters/prologix.py

```diff
@@ -86,23 +91,24 @@ def ask(self, command):
 
         :param command: SCPI command string to be sent to instrument
         """
-
+        warn("Do not call `Adapter.ask`, but `Instrument.ask` instead.",
+             FutureWarning)
         self.write(command)
         if self.rw_delay is not None:
             time.sleep(self.rw_delay)
         return self.read()
 
-    def write(self, command):
+    def _write(self, command, **kwargs):
         """ Writes the command to the GPIB address stored in the
         :attr:`.address`
 
         :param command: SCPI command string to be sent to the instrument
         """
         if self.address is not None:
-            address_command = "++addr %d\n" % self.address
-            self.connection.write(address_command.encode())
-        command += "\n"
-        self.connection.write(command.encode())
+            address_command = "++addr %d" % self.address
+            super()._write(address_command)
+            self.log.debug(("WRITE:", address_command))  # To log all communication.
```

Comment:
see above remark, this ends up being out-of-order in the log, right?

Reply:
Now I see your point. You are right. Thanks for your spotting. 

Mhmm. It's better to log after the fact has been accomplished.
We cannot know, whether the _write method does more reading/writing before or after.

Maybe we should allow to subclass the write method. Moving this logic into the write method itself solves everything.

Reply:
Yeah, I currently don't have a "great" solution, it's a tricky indirection problem.
I'm not sure if it's better to log afterwards... if communication throws an error when writing, you don't know which command triggered it all because the log message would have come _after_ the exception..

Reply:
But you know what you called.
If it succeeds, you log the result, if not, you get the exception.

Reply:
> But you know what you called.

If you're working in the REPL, sure. In a long-running application, you don't necessarily know which command put the instrument in an error state :-P (well, you maybe can infer from other log entries if your program flow is well-documented, but one shouldn't have to rely on that I think).

Reply:
You can (and probably should) put the instrument call in a try except to catch a disconnect etc, lest an instrument hickup stops your program. In this case you already handle the exception, and the handling code in turn should know, what it tried to achieve. There you might log the error (e.g."setting sweep range failed.")